TheJumpCloud/support
# Sa 2317 update jc module sdk

## Issues
* [SA-2317](https://jumpcloud.atlassian.net/browse/SA-2317) - Check for SDK updates on Connect-JCOnline

## What does this solve?
As the JumpCloud PowerShell Module and the PowerShell SDKs are more tightly intertwined, the PowerShell Module should prompt to update the SDKs when new versions are found. 

Today, if you install the PowerShell Module from PSGallery, it'll fetch the latest JCSDK modules and install them since they are required by the PowerShell Module. If new SDK versions are released after the customer installs the main PowerShell module, there is no mechanism to check for SDK updates, this PR solves for this by adding a prompt to the Update-JCModule Function.

## Is there anything particularly tricky?

Yes - [Namespace conflicts occur](https://stackoverflow.com/questions/3369662/can-you-remove-an-add-ed-type-in-powershell-again) when a newly installed module is imported on top of the existing and loaded SDK module. At a high level running the Connect-JCOnline function will run the following sequence:

1. Import Installed PowerShell Module & current JCSDK Modules
2. Set Environment Variables for API Key & Org ID 
3. Run Update-JCModule (and check for JCSDK updates + JCModule updates)

Because we generate a .dll assembly file for the JumpCloud.SDK.* Modules, the "Name" of this file in memory can conflict with other versions we attempt to import. If a SDK module is loaded in memory and a new version is loaded with `Import-Module` the PowerShell session will throw an error about the assembly file being loaded already. I believe we can solve for this in the future by appending the SDK version number to the SDK assembly name (ex. JumpCloud.SDK.V2-0.0.31 instead of JumpCloud.SDK.V2).

For now, we will prompt the user of the PowerShell terminal window that they will need to close their current session and reload in order to use the latest SDK versions:

![Screen Shot 2022-02-10 at 9 13 16 AM](https://user-images.githubusercontent.com/54448601/153448910-38048966-53f3-4844-9bd2-41f2f6e95d9d.png)

## How should this be tested?
--- Test 1 --- Test that old versions of the SDKs are updated to the latest
1. On a local system, uninstall the currently installed JumpCloud PowerShell Module `Uninstall-Module JumpCloud -Force`
2. Uninstall the SDKs `Uninstall-Module -Name "JumpCloud.SDK.V1" -Force; Uninstall-Module -Name "JumpCloud.SDK.V2" -Force; Uninstall-Module -Name "JumpCloud.SDK.DirectoryInsights" -Force`
3. Install the second latest SDK module versions `Install-Module -Name "JumpCloud.SDK.V1" -RequiredVersion "0.0.26" -Force; Install-Module -Name "JumpCloud.SDK.V2" -RequiredVersion "0.0.30" -Force; Install-Module -Name "JumpCloud.SDK.DirectoryInsights" -RequiredVersion "0.0.14" -Force`
4. Exit the session and load a new session to verify nothing is left in memory
5. Import the development version of the JumpCloud PowerShell Module from this branch `Import-Module "/Path/To/support/PowerShell/JumpCloud Module/JumpCloud.psd1" -Force`
6. Run `Update-JCModule`
7.  The following table should display and prompt to update the PowerShell Module SDKs to the latest versions
![Screen Shot 2022-02-10 at 9 17 11 AM](https://user-images.githubusercontent.com/54448601/153449657-6580794a-5eee-4ab2-b8d4-1b77d3a04e9c.png)
8. A warning message may prompt that the SDKs could not be automatically imported and to restart the PowerShell session.
9. The currently installed SDKs should be updated from their previous versions to the latest versions 
10. Verify with `Get-InstalledModule -Name JumpCloud.SDK.V1 -allVersions`, `Get-InstalledModule -Name JumpCloud.SDK.V2 -allVersions`, `Get-InstalledModule -Name JumpCloud.SDK.DirectoryInsights -allVersions`

--- Test 2 --- Test that old versions of the SDKs are uninstalled
1. On a local system, uninstall the currently installed JumpCloud PowerShell Module `Uninstall-Module JumpCloud -Force`
2. Uninstall the SDKs `Uninstall-Module -Name "JumpCloud.SDK.V1" -Force; Uninstall-Module -Name "JumpCloud.SDK.V2" -Force; Uninstall-Module -Name "JumpCloud.SDK.DirectoryInsights" -Force`
3. Install the latest SDK module versions `Install-Module -Name "JumpCloud.SDK.V1" -Force; Install-Module -Name "JumpCloud.SDK.V2" -Force; Install-Module -Name "JumpCloud.SDK.DirectoryInsights" -Force`
4. Also install an older version of the V2 module `Install-Module -Name "JumpCloud.SDK.V2" -RequiredVersion "0.0.30" -Force`
5. Exit the session and load a new session to verify nothing is left in memory
6. Import the development version of the JumpCloud PowerShell Module from this branch `Import-Module "/Path/To/support/PowerShell/JumpCloud Module/JumpCloud.psd1" -Force`
7. Run `Update-JCModule`
8.  No prompt should display but the .30 version of the SDK.V2 module should be uninstalled
9. Verify with `Get-InstalledModule -Name JumpCloud.SDK.V2 -allVersions`

--- Test 2 --- Test that old versions of the SDKs are uninstalled and only versions that need updates are installed
1. On a local system, uninstall the currently installed JumpCloud PowerShell Module `Uninstall-Module JumpCloud -Force`
2. Uninstall the SDKs `Uninstall-Module -Name "JumpCloud.SDK.V1" -Force; Uninstall-Module -Name "JumpCloud.SDK.V2" -Force; Uninstall-Module -Name "JumpCloud.SDK.DirectoryInsights" -Force`
3. Install these SDK module versions `Install-Module -Name "JumpCloud.SDK.V1 -RequiredVersion "0.0.26" -Force" -Force; Install-Module -Name "JumpCloud.SDK.V2" -Force; Install-Module -Name "JumpCloud.SDK.DirectoryInsights" -Force`
4. Also install an older version of the V2 module `Install-Module -Name "JumpCloud.SDK.V2" -RequiredVersion "0.0.30" -Force`
5. Exit the session and load a new session to verify nothing is left in memory
6. Import the development version of the JumpCloud PowerShell Module from this branch `Import-Module "/Path/To/support/PowerShell/JumpCloud Module/JumpCloud.psd1" -Force`
7. Run `Update-JCModule`
8. The prompt tabled should show that the V1 Module needs an update
9. The .30 version of the SDK.V2 module should be uninstalled
11. Verify with `Get-InstalledModule -Name JumpCloud.SDK.V2 -allVersions`,  `Get-InstalledModule -Name JumpCloud.SDK.V1 -allVersions`,  `Get-InstalledModule -Name JumpCloud.SDK.DirectoryInsights -allVersions`

## Screenshots
